src/main/kotlin/io/libp2p/etc/util/TimeLimitedMap.kt

```diff
@@ -0,0 +1,46 @@
+package io.libp2p.etc.util
+
+import java.time.Duration
+import java.util.LinkedList
+
+/**
+ * Map where entries are purged after timeout from the moment when the key was initially added.
+ *
+ * NOT thread safe
+ */
+class TimeLimitedMap<K, V>(
+    private val backMap: MutableMap<K, V>,
+    private val timeSupplier: () -> Long,
+    private val timeout: Duration
+) : MutableMap<K, V> by backMap {
+    private val timestamps = LinkedList<Pair<Long, K>>()
+
+    override fun put(key: K, value: V): V? {
+        val oldVal = backMap[key]
+
+        if (oldVal == null) {
+            timestamps += timeSupplier() to key
+        }
+        backMap[key] = value
+        purge()
```

Comment:
Should we be purging first thing?  If the timeout is 1sec, and you `put(a,1)` then 1.5 seconds later you `put(a,1)`, shouldn't the value be set with a new timeout??

Reply:
Replaced my custom class with Guava `Cache` implementation. This addresses both concerns

Reply:
Regarding my implementation: I just made the minimal simplified class which suffices `seenMessage` requirements :)

Reply:
makes sense - its just easier for long-term maintainability if classes function as expected independent of context